### DISCUSSION TITLE
Port already in use error after shutdown?

### DISCUSSION THREAD

**Discussion by willmckeandeb** [mitmproxy/mitmproxy#7237] [General]
When I shutdown then start mitmproxy within the same script, I get the following error:

`[Errno 48] HTTP(S) proxy failed to listen on *:8080 with [Errno 48] error while attempting to bind on address ('0.0.0.0', 8080): address already in use`

Code adapted from [this very useful example](https://gist.github.com/sebdelsol/2776842e655bd02b45a5be20c9aa7c09):

```python
import asyncio
import threading
from typing import Self

from mitmproxy.options import Options
from mitmproxy.tools.dump import DumpMaster


class ThreadedMitmProxy(threading.Thread):
    def __init__(self):
        self.loop = asyncio.get_event_loop()
        self.master = DumpMaster(
            Options(listen_port=8080),
            loop=self.loop,
            with_termlog=True,
            with_dumper=True,
        )
        super().__init__()

    def run(self) -> None:
        self.loop.run_until_complete(self.master.run())

    def __enter__(self) -> Self:
        self.start()
        print("started mitmproxy")
        return self

    def __exit__(self, *_) -> None:
        self.master.shutdown()
        self.join()


with ThreadedMitmProxy():
    input("waiting")
    print("shutdown mitmproxy")

with ThreadedMitmProxy():
    input("waiting")
    print("shutdown mitmproxy")
```

output:

```shell
% test_proxy.py
started mitmproxy
waiting
shutdown mitmproxy
started mitmproxy
waiting[17:58:50.363] [Errno 48] HTTP(S) proxy failed to listen on *:8080 with [Errno 48] error while attempting to bind on address ('0.0.0.0', 8080): address already in use
Try specifying a different port by using `--mode regular@8082`.
Error logged during startup, exiting...

shutdown mitmproxy
```

Any suggestions appreciated

**Comment by emanuele-em** [comment]
I think the problem could be that the os needs some time to release the port, you could try to add a short delay to allow the OS to release the port before starting the second instance 

```
with ThreadedMitmProxy():
    input("waiting")
    print("shutdown mitmproxy")
    
time.sleep(3)

with ThreadedMitmProxy():
    input("waiting")
    print("shutdown mitmproxy")
```

p.s. not tested! ⚠️

**Reply by willmckeandeb** [reply]
@emanuele-em thanks for your reply. Unfortunately the wait hasn't worked and the port/socket is still occupied. I'm not looking to use a different port for each instance. Thanks again

**Comment by lhb2** [comment]
before shutdown master,you need close server as ctx.options.update(server=False)
if shutdown from another thread,you need call with ctx.master.event_loop.call_soon_threadsafe

    def stopServer(self):
        ctx.options.update(server=False)

    ctx.master.event_loop.call_soon_threadsafe(self.stopServer)
    ctx.master.shutdown()

**Reply by willmckeandeb** [reply]
This worked perfectly - thanks for your response!
